src/components/application_manager/src/commands/command_impl.cc

```diff
@@ -130,6 +175,52 @@ void CommandImpl::ReplaceMobileByHMIAppId(
   }
 }
 
+bool CommandImpl::ReplaceHMIWithMobileAppId(
+    NsSmartDeviceLink::NsSmartObjects::SmartObject& message) {
+  if (message.keyExists(strings::app_id)) {
+    ApplicationSharedPtr application =
+        application_manager_.application_by_hmi_app(
+            message[strings::app_id].asUInt());
+
+    if (!application) {
+      LOG4CXX_ERROR(logger_, "Substitution HMI --> mobile id is failed.");
+      return false;
+    }
+    LOG4CXX_DEBUG(logger_,
+                  "ReplaceHMIWithMobileAppId from "
+                      << message[strings::app_id].asInt() << " to "
+                      << application->app_id());
+    message[strings::app_id] = application->app_id();
+  } else {
+    switch (message.getType()) {
+      case smart_objects::SmartType::SmartType_Array: {
+        smart_objects::SmartArray* message_array = message.asArray();
+        smart_objects::SmartArray::iterator it = message_array->begin();
+        for (; it != message_array->end(); ++it) {
+          if (!ReplaceHMIWithMobileAppId(*it)) {
+            break;
```

Comment:
break will cause the function return true, however it shall return false here

Reply:
@yang1070 
Done with 866dfe8

Reply:
@yang1070 
Reverted previous changes due to issues with ATF tests. As I found there is a mixed logic for two use cases: first is related to bugfix done for Ford Sync transport switch implementation which requires checking of application with app_id, second is common usage when app_id is replaced during message processing. So I propose to leave code as is for now and create another issue for investigation how this part can be re-factored.